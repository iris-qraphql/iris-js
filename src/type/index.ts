export type { Path as ResponsePath } from '../jsutils/Path';

export {
  // Predicate
  isSchema,
  // Assertion
  assertSchema,
  // GraphQL Schema definition
  GraphQLSchema,
} from './schema';
export type { GraphQLSchemaConfig, GraphQLSchemaExtensions } from './schema';

export type { GraphQLScalarType } from './definition';
export {
  isType,
  isScalarType,
  isObjectType,
  isUnionType,
  isEnumType,
  isInputObjectType,
  isListType,
  isNonNullType,
  isInputType,
  isOutputType,
  isLeafType,
  isResolverType,
  isAbstractType,
  isWrappingType,
  isNullableType,
  isNamedType,
  isRequiredArgument,
  isRequiredInputField,
  assertScalarType,
  assertObjectType,
  assertResolverType,
  assertListType,
  assertNonNullType,
  assertLeafType,
  assertCompositeType,
  assertAbstractType,
  getNullableType,
  getNamedType,
  GraphQLList,
  GraphQLNonNull,
} from './definition';

export type {
  GraphQLType,
  GraphQLInputType,
  GraphQLOutputType,
  GraphQLLeafType,
  IrisResolverType,
  GraphQLWrappingType,
  GraphQLNullableType,
  GraphQLNamedType,
  GraphQLNamedInputType,
  GraphQLNamedOutputType,
  ThunkObjMap,
  GraphQLArgument,
  GraphQLArgumentConfig,
  IrisDataVariant,
  GraphQLField,
  GraphQLFieldConfig,
  GraphQLFieldConfigArgumentMap,
  GraphQLFieldConfigMap,
  GraphQLFieldMap,
  GraphQLFieldResolver,
  GraphQLInputField,
  GraphQLIsTypeOfFn,
  GraphQLResolveInfo,
  GraphQLTypeResolver,
  GraphQLScalarSerializer,
  GraphQLScalarValueParser,
  GraphQLScalarLiteralParser,
} from './definition';

export {
  // Predicate
  isDirective,
  // Assertion
  assertDirective,
  // Directives Definition
  GraphQLDirective,
  // Built-in Directives defined by the Spec
  isSpecifiedDirective,
  specifiedDirectives,
  GraphQLIncludeDirective,
  GraphQLSkipDirective,
  GraphQLDeprecatedDirective,
  GraphQLSpecifiedByDirective,
  // Constant Deprecation Reason
  DEFAULT_DEPRECATION_REASON,
} from './directives';

export type {
  GraphQLDirectiveConfig,
  GraphQLDirectiveExtensions,
} from './directives';

// Common built-in scalar instances.
export {
  // Predicate
  isSpecifiedScalarType,
  // Standard GraphQL Scalars
  specifiedScalarTypes,
  GraphQLInt,
  GraphQLFloat,
  GraphQLString,
  GraphQLBoolean,
  GraphQLID,
  // Int boundaries constants
  GRAPHQL_MAX_INT,
  GRAPHQL_MIN_INT,
} from './scalars';

export {
  // Predicate
  isIntrospectionType,
  // GraphQL Types for introspection.
  introspectionTypes,
  __Schema,
  __Directive,
  __DirectiveLocation,
  __Type,
  __Field,
  __InputValue,
  __EnumValue,
  __TypeKind,
  // "Enum" of Type Kinds
  TypeKind,
  // Meta-field definitions.
  SchemaMetaFieldDef,
  TypeMetaFieldDef,
  TypeNameMetaFieldDef,
} from './introspection';

// Validate GraphQL schema.
export { validateSchema, assertValidSchema } from './validate';

// Upholds the spec rules about naming.
export { assertName, assertEnumValueName } from './assertName';
