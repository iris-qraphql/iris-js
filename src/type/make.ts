import type { GraphQLScalarTypeConfig } from 'graphql';
import { GraphQLScalarType } from 'graphql';

import type { IrisType } from './definition';
import { IrisTypeDefinition, IrisTypeRef } from './definition';

export const emptyDataType = (name: string) =>
  new IrisTypeDefinition({ role: 'data', name, variants: [] });

export const gqlScalar = <I, O>(config: GraphQLScalarTypeConfig<I, O>) =>
  new IrisTypeDefinition({
    role: 'data',
    name: config.name,
    description: config.description,
    scalar: new GraphQLScalarType(config),
    variants: [],
  });

export const maybe = <T extends IrisType>(ofType: T) =>
  new IrisTypeRef('MAYBE', ofType);

export const gqlList = <T extends IrisType>(ofType: T) =>
  new IrisTypeRef('LIST', ofType);
