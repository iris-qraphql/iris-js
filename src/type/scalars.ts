import type { GraphQLScalarType } from 'graphql';
import {
  GraphQLBoolean,
  GraphQLFloat,
  GraphQLID,
  GraphQLInt,
  GraphQLString,
} from 'graphql';

import type { IrisNamedType } from './definition';
import { IrisTypeDefinition } from './definition';

const liftScalar = <I, O = I>(scalar: GraphQLScalarType<I, O>) =>
  new IrisTypeDefinition({
    role: 'data',
    name: scalar.name,
    description: scalar.description,
    variants: [{ name: scalar.name }],
    scalar,
  });

export const IrisInt = liftScalar<number>(GraphQLInt);

export const IrisFloat = liftScalar<number>(GraphQLFloat);

export const IrisString = liftScalar(GraphQLString);

export const IrisBool = liftScalar(GraphQLBoolean);

export const IrisID = liftScalar(GraphQLID);

export const specifiedScalarTypes: ReadonlyArray<IrisTypeDefinition<'data'>> =
  Object.freeze([IrisString, IrisInt, IrisFloat, IrisBool, IrisID]);

export function isSpecifiedScalarType(type: IrisNamedType): boolean {
  return specifiedScalarTypes.some(({ name }) => type.name === name);
}
