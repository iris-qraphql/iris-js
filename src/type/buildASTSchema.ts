import type { ParseOptions, Source } from 'graphql';

import type { ObjMap } from '../jsutils/ObjMap';
import { keyMap } from '../jsutils/ObjMap';

import type {
  ArgumentDefinitionNode,
  DirectiveDefinitionNode,
  DocumentNode,
  FieldDefinitionNode,
  NamedTypeNode,
  ResolverVariantDefinitionNode,
  Role,
  TypeDefinitionNode,
  TypeNode,
  VariantDefinitionNode,
} from '../language/ast';
import { IrisKind } from '../language/kinds';
import { parse } from '../language/parser';
import { isTypeDefinitionNode } from '../language/predicates';

import { validateSDL } from '../validation/validate';

import { valueFromAST } from '../conversion/valueFromAST';
import { getDirectiveValues } from '../conversion/values';
import type { ConfigMap, Maybe } from '../utils/type-level';

import type {
  IrisArgument,
  IrisFieldConfig,
  IrisNamedType,
  IrisResolverVariantConfig,
  IrisType,
  IrisVariant,
} from './definition';
import { IrisDataType, IrisResolverType, IrisTypeRef } from './definition';
import { GraphQLDeprecatedDirective, GraphQLDirective } from './directives';
import { specifiedScalarTypes } from './scalars';
import type { IrisSchemaValidationOptions } from './schema';
import { IrisSchema } from './schema';

export function buildSchema(
  source: string | Source,
  options?: IrisSchemaValidationOptions & ParseOptions,
): IrisSchema {
  const document = parse(source, { noLocation: options?.noLocation });

  return buildASTSchema(document, { ...options });
}

export function buildASTSchema(
  documentAST: DocumentNode,
  options?: IrisSchemaValidationOptions,
): IrisSchema {
  if (options?.assumeValid !== true && options?.assumeValidSDL !== true) {
    const errors = validateSDL(documentAST);
    if (errors.length !== 0) {
      throw new Error(errors.map((error) => error.message).join('\n\n'));
    }
  }

  const directiveDefs: Array<DirectiveDefinitionNode> = [];
  const typeMap: Record<string, IrisNamedType> = {};

  documentAST.definitions.forEach((def) => {
    if (isTypeDefinitionNode(def)) {
      const name = def.name.value;
      typeMap[name] = stdTypeMap[name] ?? buildType(def);
    } else if (def.kind === IrisKind.DIRECTIVE_DEFINITION) {
      directiveDefs.push(def);
    }
  });

  return new IrisSchema({
    description: undefined,
    types: Object.values(typeMap),
    query: typeMap.Query as IrisResolverType,
    mutation: typeMap.Mutation as IrisResolverType,
    subscription: typeMap.Subscription as IrisResolverType,
    directives: directiveDefs.map(buildDirective),
    assumeValid: options?.assumeValid ?? false,
  });

  function getNamedType(node: ResolverVariantDefinitionNode): IrisResolverType;
  function getNamedType(node: NamedTypeNode): IrisNamedType;
  function getNamedType(
    node: NamedTypeNode | ResolverVariantDefinitionNode,
  ): IrisNamedType {
    const name = node.name.value;
    const type = stdTypeMap[name] ?? typeMap[name];

    if (type === undefined) {
      throw new Error(`Unknown type: "${name}".`);
    }
    return type;
  }

  function getWrappedType(node: TypeNode): IrisType {
    if (node.kind === IrisKind.LIST_TYPE) {
      return new IrisTypeRef('LIST', getWrappedType(node.type));
    }
    if (node.kind === IrisKind.MAYBE_TYPE) {
      return new IrisTypeRef('MAYBE', getWrappedType(node.type));
    }
    return getNamedType(node);
  }

  function buildDirective(node: DirectiveDefinitionNode): GraphQLDirective {
    return new GraphQLDirective({
      name: node.name.value,
      description: node.description?.value,
      // @ts-expect-error
      locations: node.locations.map(({ value }) => value),
      isRepeatable: node.repeatable,
      args: buildArgumentMap(node.arguments),
      astNode: node,
    });
  }

  function buildArgumentMap(
    args: Maybe<ReadonlyArray<ArgumentDefinitionNode>>,
  ): ConfigMap<IrisArgument> {
    const argsNodes = /* c8 ignore next */ args ?? [];

    const argConfigMap = Object.create(null);
    for (const arg of argsNodes) {
      // Note: While this could make assertions to get the correctly typed
      // value, that would throw immediately while type system validation
      // with validateSchema() will produce more actionable results.
      const type: any = getWrappedType(arg.type);

      argConfigMap[arg.name.value] = {
        type,
        description: arg.description?.value,
        defaultValue: valueFromAST(arg.defaultValue, type),
        deprecationReason: getDeprecationReason(arg),
        astNode: arg,
      };
    }
    return argConfigMap;
  }

  function buildFieldMap<R extends Role>(
    fields: ReadonlyArray<FieldDefinitionNode>,
  ): ObjMap<IrisFieldConfig<R>> {
    const entries = fields.map((field) => {
      const type: any = getWrappedType(field.type);
      return [
        field.name.value,
        {
          type,
          description: field.description?.value,
          deprecationReason: getDeprecationReason(field),
          astNode: field,
          args: field.arguments ? buildArgumentMap(field.arguments) : undefined,
        },
      ];
    });

    return Object.fromEntries(entries);
  }

  function buildVariant<R extends Role>(
    role: R,
    astNode: ResolverVariantDefinitionNode,
  ): IrisResolverVariantConfig | IrisVariant<'data'> {
    const name = astNode.name.value;
    const description = astNode.description?.value;
    const deprecationReason = getDeprecationReason(astNode);

    if (!astNode.fields) {
      return {
        name,
        description,
        deprecationReason,
        astNode,
        type: getNamedType(astNode),
      };
    }

    if (role === 'data') {
      return {
        name,
        description,
        deprecationReason,
        fields: astNode.fields
          ? () => buildFieldMap<'data'>(astNode?.fields ?? [])
          : undefined,
        astNode,
      } as IrisVariant<'data'>;
    }

    return {
      name,
      description,
      deprecationReason,
      fields: buildFieldMap<'resolver'>(astNode.fields ?? []),
      astNode,
    };
  }

  function buildType(astNode: TypeDefinitionNode): IrisNamedType {
    const name = astNode.name.value;

    switch (astNode.kind) {
      case IrisKind.RESOLVER_TYPE_DEFINITION: {
        return new IrisResolverType({
          name,
          description: astNode.description?.value,
          variants: () =>
            astNode.variants.map((v) => buildVariant('resolver', v)),
          astNode,
        });
      }
      case IrisKind.DATA_TYPE_DEFINITION: {
        return new IrisDataType({
          name,
          description: astNode.description?.value,
          variants: astNode.variants.map((v) => buildVariant('data', v)) as any,
          astNode,
        });
      }
    }
  }
}

const stdTypeMap = keyMap([...specifiedScalarTypes], (type) => type.name);

const getDeprecationReason = (
  node: FieldDefinitionNode | ArgumentDefinitionNode | VariantDefinitionNode,
): Maybe<string> =>
  getDirectiveValues(GraphQLDeprecatedDirective, node)?.reason as string;
