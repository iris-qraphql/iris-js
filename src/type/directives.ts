import { assertName } from 'graphql';

import type { Maybe } from '../jsutils/Maybe';

import type { DirectiveDefinitionNode } from '../language/ast';
import { DirectiveLocation } from '../language/directiveLocation';

import { inspect, instanceOf } from '../utils/legacy';
import type { ConfigMap } from '../utils/type-level';

import type { IrisArgument } from './definition';
import { buildArguments } from './definition';
import { IrisString } from './scalars';

/**
 * Test if the given value is a GraphQL directive.
 */
export function isDirective(directive: unknown): directive is GraphQLDirective {
  return instanceOf(directive, GraphQLDirective);
}

export function assertDirective(directive: unknown): GraphQLDirective {
  if (!isDirective(directive)) {
    throw new Error(
      `Expected ${inspect(directive)} to be a GraphQL directive.`,
    );
  }
  return directive;
}

/**
 * Custom extensions
 *
 * @remarks
 * Use a unique identifier name for your extension, for example the name of
 * your library or project. Do not use a shortened identifier as this increases
 * the risk of conflicts. We recommend you add at most one extension field,
 * an object which can contain all the values you need.
 */
export type GraphQLDirectiveExtensions = Record<string, unknown>;

/**
 * Directives are used by the GraphQL runtime as a way of modifying execution
 * behavior. Type system creators will usually not create these directly.
 */
export class GraphQLDirective {
  name: string;
  description: Maybe<string>;
  locations: ReadonlyArray<DirectiveLocation>;
  args: ReadonlyArray<IrisArgument>;
  isRepeatable: boolean;
  astNode: Maybe<DirectiveDefinitionNode>;

  constructor(config: Readonly<GraphQLDirectiveConfig>) {
    this.name = assertName(config.name);
    this.description = config.description;
    this.locations = config.locations;
    this.isRepeatable = config.isRepeatable ?? false;
    this.astNode = config.astNode;
    this.args = buildArguments(config.args ?? {});
  }

  get [Symbol.toStringTag]() {
    return 'GraphQLDirective';
  }

  toString(): string {
    return '@' + this.name;
  }

  toJSON(): string {
    return this.toString();
  }
}

export interface GraphQLDirectiveConfig {
  name: string;
  description?: Maybe<string>;
  locations: ReadonlyArray<DirectiveLocation>;
  args?: ConfigMap<IrisArgument>;
  isRepeatable?: Maybe<boolean>;
  astNode?: Maybe<DirectiveDefinitionNode>;
}

/**
 * Used to declare element of a GraphQL schema as deprecated.
 */
export const GraphQLDeprecatedDirective: GraphQLDirective =
  new GraphQLDirective({
    name: 'deprecated',
    description: 'Marks an element of a GraphQL schema as no longer supported.',
    locations: [
      DirectiveLocation.FIELD_DEFINITION,
      DirectiveLocation.ARGUMENT_DEFINITION,
      DirectiveLocation.INPUT_FIELD_DEFINITION,
      DirectiveLocation.ENUM_VALUE,
    ],
    args: {
      reason: {
        type: IrisString,
        description:
          'Explains why this element was deprecated, usually also including a suggestion for how to access supported similar data. Formatted using the Markdown syntax, as specified by [CommonMark](https://commonmark.org/).',
        defaultValue: '',
      },
    },
  });

/**
 * The full list of specified directives.
 */
export const specifiedDirectives: ReadonlyArray<GraphQLDirective> =
  Object.freeze([GraphQLDeprecatedDirective]);

export function isSpecifiedDirective(directive: GraphQLDirective): boolean {
  return specifiedDirectives.some(({ name }) => name === directive.name);
}
