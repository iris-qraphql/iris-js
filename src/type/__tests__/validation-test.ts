import { inspect } from '../../jsutils/inspect';

import { DirectiveLocation } from '../../language/directiveLocation';

import { buildSchema } from '../../utilities/buildASTSchema';

import { GraphQLError } from '../../error';
import { toJSONDeep } from '../../utils/toJSONDeep';
import type { ConfigMapValue } from '../../utils/type-level';

import type {
  GraphQLArgument,
  GraphQLFieldConfig,
  IrisNamedType,
  IrisStrictType,
  IrisType,
  IrisTypeRef,
} from '../definition';
import { assertDataType, assertResolverType } from '../definition';
import { assertDirective, GraphQLDirective } from '../directives';
import { gqlEnum, gqlList, gqlObject, gqlUnion, maybe } from '../make';
import { IrisString } from '../scalars';
import { IrisSchema } from '../schema';
import { validateSchema } from '../validate';

const SomeSchema = buildSchema(`
  data SomeScalar = Int

  resolver SomeObject = { f: SomeObject }

  resolver SomeUnion = SomeObject

  data SomeEnum = ONLY {}

  data SomeInputObject = { val: String }

  directive @SomeDirective on QUERY
`);

const SomeScalarType = assertDataType(SomeSchema.getType('SomeScalar'));
const SomeObjectType = assertResolverType(SomeSchema.getType('SomeObject'));
const SomeUnionType = assertResolverType(SomeSchema.getType('SomeUnion'));
const SomeEnumType = assertDataType(SomeSchema.getType('SomeEnum'));
const SomeInputObjectType = assertDataType(
  SomeSchema.getType('SomeInputObject'),
);

const SomeDirective = assertDirective(SomeSchema.getDirective('SomeDirective'));

function withModifiers<T extends IrisNamedType>(
  type: T,
): Array<T | IrisTypeRef<T | IrisTypeRef<T>>> {
  return [type, gqlList(type), maybe(type), maybe(gqlList(type))];
}

const outputTypes: ReadonlyArray<IrisType> = [
  ...withModifiers(IrisString),
  ...withModifiers(SomeScalarType),
  ...withModifiers(SomeEnumType),
  ...withModifiers(SomeObjectType),
  ...withModifiers(SomeUnionType),
];

const inputTypes: ReadonlyArray<IrisStrictType> = [
  ...withModifiers(IrisString),
  ...withModifiers(SomeScalarType),
  ...withModifiers(SomeEnumType),
  ...withModifiers(SomeInputObjectType),
];

const notInputTypes: ReadonlyArray<IrisType> = [
  ...withModifiers(SomeObjectType),
  ...withModifiers(SomeUnionType),
];

function schemaWithFieldType(type: IrisType): IrisSchema {
  return new IrisSchema({
    query: gqlObject({
      name: 'Query',
      fields: { f: { type } },
    }),
  });
}

const expectJSON = (schema: IrisSchema) =>
  expect(toJSONDeep(validateSchema(schema)));

const expectJSONEqual = (schema: IrisSchema, value: unknown) =>
  expect(toJSONDeep(validateSchema(schema))).toEqual(value);

describe('Type System: A Schema must have Object root types', () => {
  it('rejects a Schema whose query root resolver is not an Object', () => {
    const schema = buildSchema(`
      data Query = {
        test: String
      }
    `);
    expectJSONEqual(schema, [
      {
        message: 'Query root type must be Object type, it cannot be Query.',
        locations: [{ line: 2, column: 7 }],
      },
    ]);
  });

  it('rejects a Schema whose mutation type is an data type', () => {
    const schema = buildSchema(`
      resolver Query = {
        field: String
      }

      data Mutation = {
        test: String
      }
    `);
    expectJSONEqual(schema, [
      {
        message:
          'Mutation root type must be Object type if provided, it cannot be Mutation.',
        locations: [{ line: 6, column: 7 }],
      },
    ]);
  });

  it('rejects a Schema whose subscription type is an data type', () => {
    const schema = buildSchema(`
      resolver Query = {
        field: String
      }

      data Subscription = {
        test: String
      }
    `);
    expectJSONEqual(schema, [
      {
        message:
          'Subscription root type must be Object type if provided, it cannot be Subscription.',
        locations: [{ line: 6, column: 7 }],
      },
    ]);
  });

  it('rejects a Schema whose types are incorrectly typed', () => {
    const schema = new IrisSchema({
      query: SomeObjectType,
      // @ts-expect-error
      types: [{ name: 'SomeType' }, SomeDirective],
    });
    expectJSONEqual(schema, [
      {
        message: 'Expected GraphQL named type but got: { name: "SomeType" }.',
      },
      {
        message: 'Expected GraphQL named type but got: @SomeDirective.',
        locations: [{ line: 12, column: 3 }],
      },
    ]);
  });
});

describe('Type System: Objects must have fields', () => {
  it('accepts an Object type with fields object', () => {
    const schema = buildSchema(`
      resolver Query = {
        field: SomeObject
      }

      resolver SomeObject = {
        field: String
      }
    `);
    expectJSONEqual(schema, []);
  });

  it('accept an resolver with empty fields', () => {
    const schema = buildSchema(`
      resolver Query = {
        test: IncompleteObject
      }

      resolver IncompleteObject
    `);
    expectJSONEqual(schema, []);
  });

  it('rejects an Object type with incorrectly named fields', () => {
    const schema = schemaWithFieldType(
      gqlObject({
        name: 'SomeObject',
        fields: {
          __badName: { type: IrisString },
        },
      }),
    );
    expectJSONEqual(schema, [
      {
        message:
          'Name "__badName" must not begin with "__", which is reserved by GraphQL introspection.',
      },
    ]);
  });
});

describe('Type System: Fields args must be properly named', () => {
  it('accepts field args with valid names', () => {
    const schema = schemaWithFieldType(
      gqlObject({
        name: 'SomeObject',
        fields: {
          goodField: {
            type: IrisString,
            args: {
              goodArg: { type: IrisString },
            },
          },
        },
      }),
    );
    expectJSONEqual(schema, []);
  });

  it('rejects field arg with invalid names', () => {
    const schema = schemaWithFieldType(
      gqlObject({
        name: 'SomeObject',
        fields: {
          badField: {
            type: IrisString,
            args: {
              __badName: { type: IrisString },
            },
          },
        },
      }),
    );

    expectJSONEqual(schema, [
      {
        message:
          'Name "__badName" must not begin with "__", which is reserved by GraphQL introspection.',
      },
    ]);
  });
});

describe('Type System: Union types must be valid', () => {
  it('accepts a Union type with member types', () => {
    const schema = buildSchema(`
      resolver Query = {
        test: GoodUnion
      }

      resolver TypeA = {
        field: String
      }

      resolver TypeB = {
        field: String
      }

      resolver GoodUnion 
        = TypeA
        | TypeB
    `);
    expectJSONEqual(schema, []);
  });

  it('rejects a Union type with non-Object members types', () => {
    const schema = buildSchema(`
      resolver Query = {
        test: BadUnion
      }

      resolver TypeA = {
        field: String
      }

      resolver TypeB = {
        field: String
      }

      resolver BadUnion
        = TypeA
        | String
        | TypeB
        | Int
    `);

    expectJSONEqual(schema, [
      {
        message:
          'Union type BadUnion can only include Object types, it cannot include String.',
        locations: [{ line: 16, column: 11 }],
      },
      {
        message:
          'Union type BadUnion can only include Object types, it cannot include Int.',
        locations: [{ line: 18, column: 11 }],
      },
    ]);

    const badUnionMemberTypes = [
      IrisString,
      maybe(SomeObjectType),
      gqlList(SomeObjectType),
      SomeUnionType,
      SomeEnumType,
      SomeInputObjectType,
    ];
    for (const memberType of badUnionMemberTypes) {
      const badUnion = gqlUnion({
        name: 'BadUnion',
        // @ts-expect-error
        types: [memberType],
      });
      expectJSONEqual(schemaWithFieldType(badUnion), [
        {
          message:
            'Union type BadUnion can only include Object types, ' +
            `it cannot include ${inspect(memberType)}.`,
        },
      ]);
    }
  });
});

describe('Type System: Input Objects must have fields', () => {
  it('accepts an Input Object type with fields', () => {
    const schema = buildSchema(`
      resolver Query = {
        field(arg: SomeInputObject): String
      }

      data SomeInputObject = {
        field: String
      }
    `);
    expectJSONEqual(schema, []);
  });

  it('accept empty data type', () => {
    const schema = buildSchema(`
      resolver Query = {
        field(arg: SomeInputObject): String
      }

      data SomeInputObject
    `);

    expectJSONEqual(schema, []);
  });

  it('accepts an Input Object with breakable circular reference', () => {
    const schema = buildSchema(`
      resolver Query = {
        field(arg: SomeInputObject): String
      }

      data  SomeInputObject = {
        self: SomeInputObject?
        arrayOfSelf: [SomeInputObject?]?
        nonNullArrayOfSelf: [SomeInputObject?]
        nonNullArrayOfNonNullSelf: [SomeInputObject]
        intermediateSelf: AnotherInputObject
      }

      data  AnotherInputObject = {
        parent: SomeInputObject
      }
    `);

    expectJSONEqual(schema, []);
  });

  it('accept recursive data types', () => {
    const schema = buildSchema(`
      resolver Query = {
        field(arg: SomeInputObject): String
      }

      data SomeInputObject = {
        nonNullSelf: SomeInputObject
      }
    `);

    expectJSONEqual(schema, []);
  });

  it('rejects an Input Object type with incorrectly typed fields', () => {
    const schema = buildSchema(`
      resolver Query = {
        field(arg: SomeInputObject): String
      }

      resolver SomeObject = {
        field: String
      }

      resolver SomeUnion = SomeObject

      data SomeInputObject = {
        badObject: SomeObject
        badUnion: SomeUnion
        goodInputObject: SomeInputObject
      }
    `);
    expectJSONEqual(schema, [
      {
        message:
          'The type of SomeInputObject.badObject must be Input Type but got: SomeObject.',
        locations: [{ line: 13, column: 20 }],
      },
      {
        message:
          'The type of SomeInputObject.badUnion must be Input Type but got: SomeUnion.',
        locations: [{ line: 14, column: 19 }],
      },
    ]);
  });
});

describe('Type System: Enum types must be well defined', () => {
  it('rejects an Enum type with incorrectly named values', () => {
    const schema = schemaWithFieldType(gqlEnum('SomeEnum', ['__badName']));

    expectJSONEqual(schema, [
      {
        message:
          'Name "__badName" must not begin with "__", which is reserved by GraphQL introspection.',
      },
    ]);
  });
});

describe('Type System: Object fields must have output types', () => {
  function schemaWithObjectField(
    fieldConfig: GraphQLFieldConfig<unknown, unknown>,
  ): IrisSchema {
    const BadObjectType = gqlObject({
      name: 'BadObject',
      fields: {
        badField: fieldConfig,
      },
    });

    return new IrisSchema({
      query: gqlObject({
        name: 'Query',
        fields: {
          f: { type: BadObjectType },
        },
      }),
      types: [SomeObjectType],
    });
  }

  for (const type of outputTypes) {
    const typeName = inspect(type);
    it(`accepts an output type as an Object field type: ${typeName}`, () => {
      const schema = schemaWithObjectField({ type });
      expectJSONEqual(schema, []);
    });
  }

  it('rejects a non-type value as an Object field type', () => {
    // @ts-expect-error
    const schema = schemaWithObjectField({ type: Number });
    expectJSONEqual(schema, [
      {
        message:
          'The type of BadObject.badField must be Output Type but got: [function Number].',
      },
      {
        message: 'Expected GraphQL named type but got: [function Number].',
      },
    ]);
  });

  it('accept data as resolver field type', () => {
    const schema = buildSchema(`
      resolver Query = {
        field: [SomeInputObject]
      }

      data SomeInputObject = {
        field: String
      }
    `);
    expectJSONEqual(schema, []);
  });
});

describe('Type System: Arguments must have data  types', () => {
  function schemaWithArg(
    argConfig: ConfigMapValue<GraphQLArgument>,
  ): IrisSchema {
    const BadObjectType = gqlObject({
      name: 'BadObject',
      fields: {
        badField: {
          type: IrisString,
          args: {
            badArg: argConfig,
          },
        },
      },
    });

    return new IrisSchema({
      query: gqlObject({
        name: 'Query',
        fields: {
          f: { type: BadObjectType },
        },
      }),
      directives: [
        new GraphQLDirective({
          name: 'BadDirective',
          args: {
            badArg: argConfig,
          },
          locations: [DirectiveLocation.QUERY],
        }),
      ],
    });
  }

  for (const type of inputTypes) {
    const typeName = inspect(type);
    it(`accepts an data  type as a field arg type: ${typeName}`, () => {
      const schema = schemaWithArg({ type });
      expectJSONEqual(schema, []);
    });
  }

  for (const type of notInputTypes) {
    const typeStr = inspect(type);
    it(`rejects a non-input type as a field arg type: ${typeStr}`, () => {
      // @ts-expect-error
      const schema = schemaWithArg({ type });
      expectJSONEqual(schema, [
        {
          message: `The type of @BadDirective(badArg:) must be Input Type but got: ${typeStr}.`,
        },
        {
          message: `The type of BadObject.badField(badArg:) must be Input Type but got: ${typeStr}.`,
        },
      ]);
    });
  }

  it('rejects a non-type value as a field arg type', () => {
    // @ts-expect-error
    const schema = schemaWithArg({ type: Number });
    expectJSONEqual(schema, [
      {
        message:
          'The type of @BadDirective(badArg:) must be Input Type but got: [function Number].',
      },
      {
        message:
          'The type of BadObject.badField(badArg:) must be Input Type but got: [function Number].',
      },
      {
        message: 'Expected GraphQL named type but got: [function Number].',
      },
    ]);
  });

  it('rejects an required argument that is deprecated', () => {
    const schema = buildSchema(`
      directive @BadDirective(
        badArg: String @deprecated
        optionalArg: String? @deprecated
        anotherOptionalArg: String = "" @deprecated
      ) on FIELD

      resolver Query = {
        test(
          badArg: String @deprecated
          optionalArg: String? @deprecated
          anotherOptionalArg: String = "" @deprecated
        ): String
      }
    `);
    expectJSON(schema).toMatchSnapshot();
  });

  it('rejects a non-input type as a field arg with locations', () => {
    const schema = buildSchema(`
      resolver Query = {
        test(arg: SomeObject): String
      }

      resolver SomeObject = {
        foo: String
      }
    `);
    expectJSON(schema).toMatchSnapshot();
  });
});

describe('assertValidSchema', () => {
  it('do not throw on valid schemas', () => {
    const schema = buildSchema(`
      resolver Query = {
        foo: String
      }
    `);
    expect(validateSchema(schema)).toEqual([]);
  });

  it('include multiple errors into a description', () => {
    const schema = buildSchema('resolver SomeType');
    expect(validateSchema(schema)).toEqual([
      new GraphQLError('Query root type must be provided.'),
    ]);
  });
});
