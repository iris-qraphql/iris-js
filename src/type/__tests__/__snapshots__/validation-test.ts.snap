// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Type System: Arguments must have data  types rejects a non-input type as a field arg with locations 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 19,
        "line": 3,
      },
    ],
    "message": "The type of Query.test(arg:) must be Input Type but got: SomeObject.",
  },
]
`;

exports[`Type System: Arguments must have data  types rejects an required argument that is deprecated 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 24,
        "line": 3,
      },
      Object {
        "column": 17,
        "line": 3,
      },
    ],
    "message": "Required argument @BadDirective(badArg:) cannot be deprecated.",
  },
  Object {
    "locations": Array [
      Object {
        "column": 26,
        "line": 10,
      },
      Object {
        "column": 19,
        "line": 10,
      },
    ],
    "message": "Required argument Query.test(badArg:) cannot be deprecated.",
  },
]
`;
